pycuampcor deramp improvement

lijun99 opened the pull request:
This PR improves the deramp procedure in pycuampcor/dense_offset to address the issue of (linear) phase ramp along azimuth due to the Doppler Centroids. 

1) fix the pixelIdx error (thanks to @bhawkins)
2) use double precision for phase correction. Using float seems to introduce spectral noises which breaks the band-limited structure and therefore lowers the correlation. 
3) add derampAxis parameter for options to deramp along certain direction, 0=azimuth 1=range and 2=both axes(default). In cfg, use deramp_axis =  "azimuth", "range", or None to set this parameter.
4) move cuDeramp from oversampler to chunk processor; making the code more readable

bhawkins committed:
Use consistent X and Y indexing.

bhawkins committed:
Ditto in CPU version.

bhawkins committed:
update golden data

Commit:
deramp improvement

 1) fix the pixelIdx error (thanks to @bhawkins)
 2) use double precision for phase correction
 3) add derampAxis parameter for options to deramp along certain
    direction
 4) move cuDeramp from oversampler to chunk processor; making the code more readable

xhuang-jpl commented:
Thanks @lijun99 and @bhawkins, I will take a test.

Commit:
update the runconfig

Commit:
add the complex_no_deramp option

xhuang-jpl reviewed (approved):
I have tested this PR on both CPU and GPU, which show significant improvements of the correlation surface peak after the ramp phase estimation bug fix. Thanks again @lijun99  !

bhawkins commented on cxx/isce3/cuda/matchtemplate/pycuampcor/cuAmpcorUtil.h:
```diff
@@ -46,8 +46,8 @@ void cuArraysC2R(cuArrays<float2> *image1, cuArrays<float> *image2, cudaStream_t
 void cuArraysAbs(cuArrays<float2> *image1, cuArrays<float> *image2, cudaStream_t stream);
 
 // cuDeramp.cu: deramping phase
```
```suggestion
// cuDeramp.cu: deramping phase
// `cuDeramp` calls a deramp implementation (or does nothing) based on the value of `method`:
//  `method=1` for cuLinearDeramp, any other value for no-op
// `cuLinearDeramp` Estimates the phase gradient over the chip and removes it.
```
Some explanatory comments would be helpful.

bhawkins commented on cxx/isce3/cuda/matchtemplate/pycuampcor/cuDeramp.cu:
```diff
@@ -45,72 +45,87 @@ __device__ void complexSumReduceBlock(float2& sum, volatile float *shmem)
+        // use double to improve accuracy
+        double phase = pixelIdxX*phaseX + pixelIdxY*phaseY;
+        double phase_cos = cos(phase);
+        double phase_sin = sin(phase);
```
```suggestion
        double phase_sin, phase_cos;
        sincos(phase, &phase_sin, &phase_cos);
```
Hopefully the compiler does this transformation already, but we may as well do it explicitly ([reference](https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH__DOUBLE.html#group__cuda__math__double_1gabe0e6a063a8f38850b0323933cf3320b)).

bhawkins commented on cxx/isce3/cuda/matchtemplate/pycuampcor/cuDeramp.cu:
```diff
@@ -120,38 +135,38 @@ __global__ void cuDerampMethod1_kernel(float2 *images, const int imageNX, int co
+void cuLinearDeramp(cuArrays<float2> *images, const int axis, cudaStream_t stream)
 {
-    
+
```
```suggestion
    if ((axis < 0) or (axis > 2)) {
        throw std::invalid_argument("deramp axis must be 0, 1, or 2");
    }
```
It'd be nice to validate this user-defined input (here or earlier in the program).

bhawkins commented on share/nisar/defaults/insar.yaml:
```diff
@@ -405,8 +405,10 @@ runconfig:
                 # prior to cross-correlation computation
                 slc_oversampling_factor: 2
-                # Deramp data prior to FFT: magnitude or complex (linear phase ramp)
+                # Deramp data prior to FFT: magnitude , complex (linear phase ramp), or complex without deramp
```
```suggestion
                # Deramp data prior to FFT: "magnitude", "complex" (linear phase ramp), or "complex_no_deramp"
```
It'd be nice to have the actual valid strings in the comment here.

bhawkins commented on python/packages/nisar/workflows/offsets_product.py:
```diff
@@ -256,8 +256,25 @@ def set_ampcor_params(cfg, ampcor_obj):
+        elif deramp_axis == "range":
+            ampcor_obj.derampAxis = 1
+        else: # both directions
+            ampcor_obj.derampAxis = 2
```
```suggestion
        elif deramp_axis == "both":
            ampcor_obj.derampAxis = 2
        else:
            raise ValueError(f"invalid {deramp_axis=}")
```
It wouldn't hurt to be more paranoid. Ditto for `deramping_method`. But not a big deal.

bhawkins commented on share/nisar/defaults/insar.yaml:
```diff
@@ -456,8 +458,10 @@ runconfig:
                 # Cross-correlation domain
                 cross_correlation_domain: 'frequency'
-                # Deramp data prior to FFT: magnitude or complex (linear phase ramp)
+                # Deramp data prior to FFT: magnitude , complex (linear phase ramp), or complex without deramp
```
Ditto spelling out the valid values.

bhawkins reviewed (approved):
LGTM. I see the deramp phase calculations are double precision, and the axes are consistent between the estimation and removal steps.

I found a few things that could be tweaked but no blocking issues.

hfattahi commented:
@lijun99 would you please sync back to develop branch. There are some GDAL related failures unrelated to this PR that has been hopefully already fixed in develop.

lijun99 committed:
Apply suggestions from code review

from @bhawkins :
  - commenting and value checks on deramp_method and deramp_axis
  - use explicit sincos for potential performance boost

Co-authored-by: Brian Hawkins <1729052+bhawkins@users.noreply.github.com>

Commit:
use explicit sincos for ampcor cpu module as well

Commit:
add comments on deramp_method in runconfig

lijun99 commented:
> LGTM. I see the deramp phase calculations are double precision, and the axes are consistent between the estimation and removal steps.
> 
> I found a few things that could be tweaked but no blocking issues.

Thank you, @bhawkins !

lijun99 commented:
> @lijun99 would you please sync back to develop branch. There are some GDAL related failures unrelated to this PR that has been hopefully already fixed in develop.

Are you referring to the GDAL mmap driver crash when the file size is huge (exceeding the memory size)? Maybe we should use a different PR to resolve this. My fix was to use a home-cooked mmap cache manager. Maybe I can take a look at the H5 Raster driver to see whether I can patch there.  

BTW, what is the decision on Pycuampor V2 #76 , with two workflows? I just hope not to diverge the code development.

hfattahi commented:
> > @lijun99 would you please sync back to develop branch. There are some GDAL related failures unrelated to this PR that has been hopefully already fixed in develop.
> 
> Are you referring to the GDAL mmap driver crash when the file size is huge (exceeding the memory size)? Maybe we should use a different PR to resolve this. My fix was to use a home-cooked mmap cache manager. Maybe I can take a look at the H5 Raster driver to see whether I can patch there.
> 
> BTW, what is the decision on Pycuampor V2 #76 , with two workflows? I just hope not to diverge the code development.

No I was referring to this PR https://github.com/isce-framework/isce3/pull/196/files which is already merged.
For PyCuAmpcor V2, let's discuss separately.

Commit:
add math.h for sincos

lijun99 commented:
> > > @lijun99 would you please sync back to develop branch. There are some GDAL related failures unrelated to this PR that has been hopefully already fixed in develop.
> > 
> > 
> > Are you referring to the GDAL mmap driver crash when the file size is huge (exceeding the memory size)? Maybe we should use a different PR to resolve this. My fix was to use a home-cooked mmap cache manager. Maybe I can take a look at the H5 Raster driver to see whether I can patch there.
> > BTW, what is the decision on Pycuampor V2 #76 , with two workflows? I just hope not to diverge the code development.
> 
> No I was referring to this PR https://github.com/isce-framework/isce3/pull/196/files which is already merged. For PyCuAmpcor V2, let's discuss separately.

I am not sure what types of files are fed into pycuamcpor, slc or h5? @xhuang-jpl, could you give me some details?

hfattahi commented:
> > > > @lijun99 would you please sync back to develop branch. There are some GDAL related failures unrelated to this PR that has been hopefully already fixed in develop.
> > > 
> > > 
> > > Are you referring to the GDAL mmap driver crash when the file size is huge (exceeding the memory size)? Maybe we should use a different PR to resolve this. My fix was to use a home-cooked mmap cache manager. Maybe I can take a look at the H5 Raster driver to see whether I can patch there.
> > > BTW, what is the decision on Pycuampor V2 #76 , with two workflows? I just hope not to diverge the code development.
> > 
> > 
> > No I was referring to this PR https://github.com/isce-framework/isce3/pull/196/files which is already merged. For PyCuAmpcor V2, let's discuss separately.
> 
> I am not sure what types of files are fed into pycuamcpor, slc or h5? @xhuang-jpl, could you give me some details?

As I said it was not related to this PR. geocodeCov was failing. Syncing with develop should solve it hopefully.

lijun99 commented:
I see. develop merged.

I have tried Linux with cuda on my machine and it compiles without an issue. However, the ci tests seem to fail probably due to the different results. Could someone help to update it? Thanks! @bhawkins @xhuang-jpl

hfattahi commented:
@lijun99 would you please take a look at this unit test (https://github.com/isce-framework/isce3/blob/develop/tests/python/packages/isce3/matchtemplate/test_ampcor.py). With this PR the test is failing with the following message:

```
Processing chunks (10, x) - (10, x) out of 13
Processing chunks (11, x) - (11, x) out of 13
Processing chunks (12, x) - (12, x) out of 13
Processing chunks (13, x) - (13, x) out of 13
comparing covariance
meandiff 6.768423e-10
comparing dense_offsets
got 0.015625 but expected -0.109375 diff is 0.125
at index 18
=========================== short test summary info ============================
FAILED ../../../../../../tests/python/packages/isce3/matchtemplate/test_ampcor.py::test_ampcor - assert False
```

I wonder if the test also fails for you?

lijun99 commented:
I guess we need to update the golden data. I did try to update, and it seems that cpu and gpu results diff from the correlation_peak by a max diff 0.015. I will do more tests tomorrow to figure out whether it can be reconciled.

Tyler-g-hudson commented:
@lijun99 please also rebase to the latest develop commit in order to fix the build-and-run test failures you're getting

EDIT: Nevermind, I misunderstood the problem

Commit:
update ampcor test golden data and adjust the tolerance for correlation peak

lijun99 commented:
I did a rebase instead of merge from the current develop branch, which is the force-push about. 

I checked the ampcor test, and updated dense_offset and correlation_peak golden data. The dense_offset results were obtained from the a "faulty" procedure, which deviate from the new results by a max 0.2 (pixel). From debug runs, I checked the detailed outputs from each procedure and I think the new results are more reasonable. 

CPU and GPU results are consistent with only tiny differences ~ 0.002 pixel in offset, ~0.002 in peak values. The test also compares fft and sinc oversamplers (for the correlation surface). and there is a ~0.013 difference in peak values. I think it is within reasonable numeric errors between fft and sinc interpolation methods and therefore I adjusted peak_value tolerance to 2e-2. Now the ampcor test passes. 

Please double check.

hfattahi commented:
> I did a rebase instead of merge from the current develop branch, which is the force-push about.
> 
> I checked the ampcor test, and updated dense_offset and correlation_peak golden data. The dense_offset results were obtained from the a "faulty" procedure, which deviate from the new results by a max 0.2 (pixel). From debug runs, I checked the detailed outputs from each procedure and I think the new results are more reasonable.
> 
> CPU and GPU results are consistent with only tiny differences ~ 0.002 pixel in offset, ~0.002 in peak values. The test also compares fft and sinc oversamplers (for the correlation surface). and there is a ~0.013 difference in peak values. I think it is within reasonable numeric errors between fft and sinc interpolation methods and therefore I adjusted peak_value tolerance to 2e-2. Now the ampcor test passes.
> 
> Please double check.

Thanks @lijun99 . Let me run the tests again and after will look at your test modifications

lijun99 commented:
It looks like macOS doesn't recognize sincos. Let me workout a solution.

bhawkins commented:
> It looks like macOS doesn't recognize sincos. Let me workout a solution.

Yeah, `sincos` isn't part of C or C++ standards, so I wouldn't recommend including it in the CPU code. (It is part of the CUDA API, so fair game there.) Fortunately most C++ compilers support it as an intrinsic function and are smart enough to use it. For example, both GCC and clang do the transformation when you compile with `-ffast-math` ([godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:47,endLineNumber:10,positionColumn:47,positionLineNumber:10,selectionStartColumn:47,selectionStartLineNumber:10,startColumn:47,startLineNumber:10),source:'%23include+%3Ccstdio%3E%0A%23include+%3Ccmath%3E%0A%0Aint+main()+%7B%0A++++double+phi%3B%0A++++puts(%22enter+phase:+%22)%3B+//+avoid+optimizing+out+the+value%0A++++scanf(%22%25lf%22,+%26phi)%3B%0A%0A++++double+s+%3D+std::sin(phi),+c+%3D+std::cos(phi)%3B%0A++++printf(%22phi%3D%25g+cos%3D%25g+sin%3D%25g%5Cn%22,+phi,+c,+s)%3B%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B20+-O3+-ffast-math',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),k:35.35095197499289,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),k:31.31571469167377,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)).

Commit:
fix sincos for macOS

lijun99 commented:
> > It looks like macOS doesn't recognize sincos. Let me workout a solution.
> 
> Yeah, `sincos` isn't part of C or C++ standards, so I wouldn't recommend including it in the CPU code. (It is part of the CUDA API, so fair game there.) Fortunately most C++ compilers support it as an intrinsic function and are smart enough to use it. For example, both GCC and clang do the transformation when you compile with `-ffast-math` ([godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:47,endLineNumber:10,positionColumn:47,positionLineNumber:10,selectionStartColumn:47,selectionStartLineNumber:10,startColumn:47,startLineNumber:10),source:'%23include+%3Ccstdio%3E%0A%23include+%3Ccmath%3E%0A%0Aint+main()+%7B%0A++++double+phi%3B%0A++++puts(%22enter+phase:+%22)%3B+//+avoid+optimizing+out+the+value%0A++++scanf(%22%25lf%22,+%26phi)%3B%0A%0A++++double+s+%3D+std::sin(phi),+c+%3D+std::cos(phi)%3B%0A++++printf(%22phi%3D%25g+cos%3D%25g+sin%3D%25g%5Cn%22,+phi,+c,+s)%3B%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B20+-O3+-ffast-math',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),k:35.35095197499289,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),k:31.31571469167377,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)).

I tried __sincos in macOS and it indeed works. So, which do you prefer, a __APPLE__ check for __sincos, as I just pushed, or resort to sin/cos separate lines?

bhawkins commented:
> I tried __sincos in macOS and it indeed works. So, which do you prefer, a **APPLE** check for __sincos, as I just pushed, or resort to sin/cos separate lines?

That's good to know. However, isce3 is currently aiming for C++17 language standard compliance, so I'd prefer to avoid explicit `sincos` and rely on the compiler to optimize it.

Commit:
revert sincos in the cpu module to separate calls (optimize by  -ffast-math)

hfattahi reviewed (approved):
LGTM. Thanks @lijun99 for these improvements. We have tested extensively thanks to @xhuang-jpl for testing and verifying the fixes.
